Fix flakiness in least squares test

This test is still failing occasionally. Needs robustification,

Here's an example failure:

24/57 Testing: test_least_squares

24/57 Test: test_least_squares

Command: "/home/travis/build/polymec/polymec-dev/build/Linux-x86_64-static-double-cc-Debug/core/tests/test_least_squares"

Directory: /home/travis/build/polymec/polymec-dev/build/Linux-x86_64-static-double-cc-Debug/core/tests

"test_least_squares" start time: Feb 22 04:48 UTC

Output:

----------------------------------------------------------

[==========] Running 8 test(s).

[ RUN      ] test_p0_fit

[       OK ] test_p0_fit

[ RUN      ] test_p1_fit

[       OK ] test_p1_fit

[ RUN      ] test_p2_fit

[       OK ] test_p2_fit

[ RUN      ] test_p3_fit

[       OK ] test_p3_fit

[ RUN      ] test_weighted_p0_fit

[       OK ] test_weighted_p0_fit

[ RUN      ] test_weighted_p1_fit

[       OK ] test_weighted_p1_fit

[ RUN      ] test_weighted_p2_fit

[       OK ] test_weighted_p2_fit

[ RUN      ] test_weighted_p3_fit

[  FAILED  ] test_weighted_p3_fit

[==========] 8 test(s) run.

ABS(fitted_value - actual_value) < 5e-12

/home/travis/build/polymec/polymec-dev/core/tests/test_least_squares.c:151: error: Failure!

[  PASSED  ] 7 test(s).

[  FAILED  ] 1 test(s), listed below:

[  FAILED  ] test_weighted_p3_fit

 1 FAILED TEST(S)

<end of output>

Test time =   0.00 sec

----------------------------------------------------------

Test Failed.

"test_least_squares" end time: Feb 22 04:48 UTC

"test_least_squares" time elapsed: 00:00:00

----------------------------------------------------------

--- a/core/CMakeLists.txt
+++ b/core/CMakeLists.txt
@@ -6,7 +6,7 @@ configure_file(
 
 # Certain tools are available to us only in a shared lib environment.
 if (BUILD_SHARED_LIBS)
-  set(DYLOAD_SOURCES krylov_solver.c petsc_krylov_solver.c hypre_krylov_solver.c)
+  set(DYLOAD_SOURCES petsc_krylov_solver.c hypre_krylov_solver.c)
 endif()
 
 include(add_polymec_library)
@@ -19,7 +19,8 @@ add_polymec_library(polymec_core polymec.c arch.c allocators.c logging.c timer.c
                     text_buffer.c serializer.c thread_pool.c string_utils.c 
                     file_utils.c array_utils.c lookup1.c linear_algebra.c 
                     local_matrix.c block_diagonal_matrix.c sparse_local_matrix.c 
-                    dense_local_matrix ${DYLOAD_SOURCES} silo_file.c exchanger.c)
+                    dense_local_matrix.c krylov_solver.c 
+                    ${DYLOAD_SOURCES} silo_file.c exchanger.c)
 
 # polymec.c depends on the creation of polymec_version.h.
 add_dependencies(polymec_core update_version_h)

--- a/core/krylov_solver.h
+++ b/core/krylov_solver.h
@@ -193,7 +193,7 @@ krylov_solver_t* krylov_factory_gmres_solver(krylov_factory_t* factory,
                                              MPI_Comm comm,
                                              int krylov_dimension);
                                              
-// Constructs a stabilized bi-conjugate Krylov solver.
+// Constructs a stabilized bi-conjugate gradient Krylov solver.
 krylov_solver_t* krylov_factory_bicgstab_solver(krylov_factory_t* factory,
                                                 MPI_Comm comm);
 

--- a/core/silo_file.c
+++ b/core/silo_file.c
@@ -877,12 +877,7 @@ silo_file_t* silo_file_new(MPI_Comm comm,
     char group_dir_name[FILENAME_MAX];
     snprintf(group_dir_name, FILENAME_MAX, "%s/%d", file->directory, file->group_rank);
     if (file->rank_in_group == 0)
-    {
       create_directory(group_dir_name, S_IRWXU | S_IRWXG);
-      MPI_Barrier(file->comm);
-    }
-    else
-      MPI_Barrier(file->comm);
 
     // Determine a file name and directory name.
     if (cycle == -1)
@@ -1070,10 +1065,7 @@ silo_file_t* silo_file_open(MPI_Comm comm,
       }
       else
         closedir(group_dir);
-      MPI_Barrier(file->comm);
     }
-    else
-      MPI_Barrier(file->comm);
 
     // Determine a file name and directory name.
     if (cycle == -1)

--- a/core/tests/test_least_squares.c
+++ b/core/tests/test_least_squares.c
@@ -93,7 +93,7 @@ static ls_weight_func_t* simple_w_new(int A, real_t B)
   return ls_weight_func_new("Simple", W_data, vtable);
 }
 
-void test_poly_fit(void** state, int p, point_t* x0, point_t* points, int num_points, real_t* coeffs, bool weighted)
+bool test_poly_fit(void** state, int p, point_t* x0, point_t* points, int num_points, real_t* coeffs, bool weighted)
 {
   polynomial_t* poly = polynomial_new(p, coeffs, x0);
 
@@ -147,100 +147,173 @@ void test_poly_fit(void** state, int p, point_t* x0, point_t* points, int num_po
   real_t actual_value = polynomial_value(poly, &x_star);
   polynomial_t* fitted_poly = polynomial_new(p, b, x0);
   real_t fitted_value = polynomial_value(fitted_poly, &x_star);
+  real_t error = ABS(fitted_value - actual_value);
+  log_debug("error for order %d fit at x = (%g, %g, %g) with %d points: %g", 
+            p, x_star.x, x_star.y, x_star.z, num_points, error); 
+  bool passed;
   if (weighted)
-    assert_true(ABS(fitted_value - actual_value) < 5e-12);
+    passed = (error < 5e-12);
   else
-    assert_true(ABS(fitted_value - actual_value) < 1e-11);
+    passed = (error < 1e-11);
 
   poly = NULL;
   fitted_poly = NULL;
+
+  return passed;
 }
+
+// We want 99% reliability for randomly selected points.
 
 void test_p0_fit(void** state)
 {
-  static real_t coeffs[] = {1.0};
-  point_t x0, points[4];
-  generate_random_points(4, points);
-  test_poly_fit(state, 0, NULL, points, 4, coeffs, false);
-  average_points(points, 4, &x0);
-  test_poly_fit(state, 0, &x0, points, 4, coeffs, false);
+  int num_failures = 0;
+  for (int i = 0; i < 100; ++i) 
+  {
+    static real_t coeffs[] = {1.0};
+    point_t x0, points[4];
+    generate_random_points(4, points);
+    bool passed = test_poly_fit(state, 0, NULL, points, 4, coeffs, false);
+    average_points(points, 4, &x0);
+    bool passed_x0 = test_poly_fit(state, 0, &x0, points, 4, coeffs, false);
+
+    if (!passed || !passed_x0)
+      ++num_failures;
+  }
+  assert_true(num_failures <= 1);
 }
 
 void test_weighted_p0_fit(void** state)
 {
-  static real_t coeffs[] = {1.0};
-  point_t x0, points[4];
-  generate_random_points(4, points);
-  test_poly_fit(state, 0, NULL, points, 4, coeffs, true);
-  average_points(points, 4, &x0);
-  test_poly_fit(state, 0, &x0, points, 4, coeffs, true);
+  int num_failures = 0;
+  for (int i = 0; i < 100; ++i) 
+  {
+    static real_t coeffs[] = {1.0};
+    point_t x0, points[4];
+    generate_random_points(4, points);
+    bool passed = test_poly_fit(state, 0, NULL, points, 4, coeffs, true);
+    average_points(points, 4, &x0);
+    bool passed_x0 = test_poly_fit(state, 0, &x0, points, 4, coeffs, true);
+
+    if (!passed || !passed_x0)
+      ++num_failures;
+  }
+  assert_true(num_failures <= 1);
 }
 
 void test_p1_fit(void** state)
 {
-  static real_t coeffs[] = {1.0, 2.0, 3.0, 4.0};
-  point_t x0, points[8];
-  generate_random_points(8, points);
-  test_poly_fit(state, 1, NULL, points, 8, coeffs, false);
-  average_points(points, 8, &x0);
-  test_poly_fit(state, 1, &x0, points, 8, coeffs, false);
+  int num_failures = 0;
+  for (int i = 0; i < 100; ++i) 
+  {
+    static real_t coeffs[] = {1.0, 2.0, 3.0, 4.0};
+    point_t x0, points[8];
+    generate_random_points(8, points);
+    bool passed = test_poly_fit(state, 1, NULL, points, 8, coeffs, false);
+    average_points(points, 8, &x0);
+    bool passed_x0 = test_poly_fit(state, 1, &x0, points, 8, coeffs, false);
+
+    if (!passed || !passed_x0)
+      ++num_failures;
+  }
+  assert_true(num_failures <= 1);
 }
 
 void test_weighted_p1_fit(void** state)
 {
-  static real_t coeffs[] = {1.0, 2.0, 3.0, 4.0};
-  point_t x0, points[8];
-  generate_random_points(8, points);
-  test_poly_fit(state, 1, NULL, points, 8, coeffs, true);
-  average_points(points, 8, &x0);
-  test_poly_fit(state, 1, &x0, points, 8, coeffs, true);
+  int num_failures = 0;
+  for (int i = 0; i < 100; ++i) 
+  {
+    static real_t coeffs[] = {1.0, 2.0, 3.0, 4.0};
+    point_t x0, points[8];
+    generate_random_points(8, points);
+    bool passed = test_poly_fit(state, 1, NULL, points, 8, coeffs, true);
+    average_points(points, 8, &x0);
+    bool passed_x0 = test_poly_fit(state, 1, &x0, points, 8, coeffs, true);
+
+    if (!passed || !passed_x0)
+      ++num_failures;
+  }
+  assert_true(num_failures <= 1);
 }
 
 void test_p2_fit(void** state)
 {
-  static real_t coeffs[] = {1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0};
-  point_t x0, points[16];
-  generate_random_points(16, points);
-  test_poly_fit(state, 2, NULL, points, 16, coeffs, false);
-  average_points(points, 16, &x0);
-  test_poly_fit(state, 2, &x0, points, 16, coeffs, false);
+  int num_failures = 0;
+  for (int i = 0; i < 100; ++i) 
+  {
+    static real_t coeffs[] = {1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0};
+    point_t x0, points[16];
+    generate_random_points(16, points);
+    bool passed = test_poly_fit(state, 2, NULL, points, 16, coeffs, false);
+    average_points(points, 16, &x0);
+    bool passed_x0 = test_poly_fit(state, 2, &x0, points, 16, coeffs, false);
+
+    if (!passed || !passed_x0)
+      ++num_failures;
+  }
+  assert_true(num_failures <= 1);
 }
 
 void test_weighted_p2_fit(void** state)
 {
-  static real_t coeffs[] = {1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0};
-  point_t x0, points[16];
-  generate_random_points(16, points);
-  test_poly_fit(state, 2, NULL, points, 16, coeffs, true);
-  average_points(points, 16, &x0);
-  test_poly_fit(state, 2, &x0, points, 16, coeffs, true);
+  int num_failures = 0;
+  for (int i = 0; i < 100; ++i) 
+  {
+    static real_t coeffs[] = {1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0};
+    point_t x0, points[16];
+    generate_random_points(16, points);
+    bool passed = test_poly_fit(state, 2, NULL, points, 16, coeffs, true);
+    average_points(points, 16, &x0);
+    bool passed_x0 = test_poly_fit(state, 2, &x0, points, 16, coeffs, true);
+
+    if (!passed || !passed_x0)
+      ++num_failures;
+  }
+  assert_true(num_failures <= 1);
 }
 
 void test_p3_fit(void** state)
 {
-  static real_t coeffs[] = {1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0, 
-                            11.0, 12.0, 13.0, 14.0, 15.0, 16.0, 17.0, 18.0, 19.0, 20.0};
-  point_t x0, points[30];
-  generate_random_points(30, points);
-  test_poly_fit(state, 3, NULL, points, 30, coeffs, false);
-  average_points(points, 30, &x0);
-  test_poly_fit(state, 3, &x0, points, 30, coeffs, false);
+  int num_failures = 0;
+  for (int i = 0; i < 100; ++i) 
+  {
+    static real_t coeffs[] = {1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0, 
+      11.0, 12.0, 13.0, 14.0, 15.0, 16.0, 17.0, 18.0, 19.0, 20.0};
+    point_t x0, points[40];
+    generate_random_points(40, points);
+    bool passed = test_poly_fit(state, 3, NULL, points, 40, coeffs, false);
+    average_points(points, 40, &x0);
+    bool passed_x0 = test_poly_fit(state, 3, &x0, points, 40, coeffs, false);
+
+    if (!passed || !passed_x0)
+      ++num_failures;
+  }
+  assert_true(num_failures <= 1);
 }
 
 void test_weighted_p3_fit(void** state)
 {
-  static real_t coeffs[] = {1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0, 
-                            11.0, 12.0, 13.0, 14.0, 15.0, 16.0, 17.0, 18.0, 19.0, 20.0};
-  point_t x0, points[30];
-  generate_random_points(30, points);
-  test_poly_fit(state, 3, NULL, points, 30, coeffs, true);
-  average_points(points, 30, &x0);
-  test_poly_fit(state, 3, &x0, points, 30, coeffs, true);
+  int num_failures = 0;
+  for (int i = 0; i < 100; ++i) 
+  {
+    static real_t coeffs[] = {1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0, 
+      11.0, 12.0, 13.0, 14.0, 15.0, 16.0, 17.0, 18.0, 19.0, 20.0};
+    point_t x0, points[50];
+    generate_random_points(50, points);
+    bool passed = test_poly_fit(state, 3, NULL, points, 50, coeffs, true);
+    average_points(points, 50, &x0);
+    bool passed_x0 = test_poly_fit(state, 3, &x0, points, 50, coeffs, true);
+
+    if (!passed || !passed_x0)
+      ++num_failures;
+  }
+  assert_true(num_failures <= 1);
 }
 
 int main(int argc, char* argv[]) 
 {
   polymec_init(argc, argv);
+  set_log_level(LOG_DEBUG);
 
   // Initialize the random number generator.
   srand((unsigned)time(NULL));

--- a/integrators/bdf_ode_integrator.c
+++ b/integrators/bdf_ode_integrator.c
@@ -46,6 +46,16 @@ typedef struct
   newton_pc_t* precond;
   int (*Jy)(void* context, real_t t, real_t* x, real_t* rhstx, real_t* y, real_t* temp, real_t* Jy);
 
+  // INK stuff.
+  int (*k_rhs_func)(void* context, real_t t, krylov_vector_t* x, krylov_vector_t* rhs);
+  int (*k_J_func)(void* context, real_t t, krylov_vector_t* x, krylov_vector_t* rhs, krylov_matrix_t* J);
+  krylov_solver_t* k_solver;
+  krylov_pc_t* k_precond;
+  krylov_matrix_t* k_J;
+  krylov_matrix_t* k_old_J;
+  krylov_vector_t* k_x;
+  krylov_vector_t* k_b;
+
   // Error weight function.
   void (*compute_weights)(void* context, real_t* y, real_t* weights);
   real_t* error_weights;
@@ -235,6 +245,20 @@ static void bdf_dtor(void* context)
   N_VDestroy(integ->x);
   CVodeFree(&integ->cvode);
 
+  // Kill the Inexact Newton-Krylov stuff.
+  if (integ->k_solver != NULL)
+    krylov_solver_free(integ->k_solver);
+  if (integ->k_precond != NULL)
+    krylov_pc_free(integ->k_precond);
+  if (integ->k_J != NULL)
+    krylov_matrix_free(integ->k_J);
+  if (integ->k_old_J != NULL)
+    krylov_matrix_free(integ->k_old_J);
+  if (integ->k_x != NULL)
+    krylov_vector_free(integ->k_x);
+  if (integ->k_b != NULL)
+    krylov_vector_free(integ->k_b);
+
   // Kill the rest.
   if (integ->status_message != NULL)
     polymec_free(integ->status_message);
@@ -315,19 +339,19 @@ static int eval_Jy(N_Vector y, N_Vector Jy, real_t t, N_Vector x, N_Vector rhs, 
 }
 
 ode_integrator_t* jfnk_bdf_ode_integrator_new(int order,
-                                             MPI_Comm comm,
-                                             int num_local_values, 
-                                             int num_remote_values, 
-                                             void* context, 
-                                             int (*rhs_func)(void* context, real_t t, real_t* x, real_t* xdot),
-                                             int (*Jy_func)(void* context, real_t t, real_t* x, real_t* rhs, real_t* y, real_t* temp, real_t* Jy),
-                                             void (*dtor)(void* context),
-                                             newton_pc_t* precond,
-                                             jfnk_bdf_krylov_t solver_type,
-                                             int max_krylov_dim)
+                                              MPI_Comm comm,
+                                              int num_local_values, 
+                                              int num_remote_values, 
+                                              void* context, 
+                                              int (*rhs_func)(void* context, real_t t, real_t* x, real_t* xdot),
+                                              int (*Jy_func)(void* context, real_t t, real_t* x, real_t* rhs, real_t* y, real_t* temp, real_t* Jy),
+                                              void (*dtor)(void* context),
+                                              newton_pc_t* precond,
+                                              jfnk_bdf_krylov_t solver_type,
+                                              int max_krylov_dim)
 {
   ASSERT(order >= 1);
-  ASSERT(order <= 12);
+  ASSERT(order <= 5);
   ASSERT(num_local_values > 0);
   ASSERT(num_remote_values >= 0);
   ASSERT(rhs_func != NULL);
@@ -348,6 +372,13 @@ ode_integrator_t* jfnk_bdf_ode_integrator_new(int order,
   integ->t = 0.0;
   integ->observers = ptr_array_new();
   integ->error_weights = NULL;
+
+  integ->k_rhs_func = NULL;
+  integ->k_J_func = NULL;
+  integ->k_solver = NULL;
+  integ->k_precond = NULL;
+  integ->k_J = integ->k_old_J = NULL;
+  integ->k_x = integ->k_b = NULL;
 
   // Set up KINSol and accessories.
   integ->x = N_VNew(integ->comm, integ->num_local_values);
@@ -610,5 +641,85 @@ void* bdf_ode_integrator_cvode(ode_integrator_t* integrator)
 int bdf_ode_integrator_rhs(real_t t, N_Vector x, N_Vector x_dot, void* context)
 {
   return bdf_evaluate_rhs(t, x, x_dot, context);
+}
+
+//------------------------------------------------------------------------
+//                  Inexact Newton-Krylov integrator stuff
+//------------------------------------------------------------------------
+static int ink_rhs(void* context, real_t t, real_t* x, real_t* xdot)
+{
+  return 0;
+}
+
+ode_integrator_t* ink_bdf_ode_integrator_new(int order, 
+                                             MPI_Comm comm,
+                                             int num_local_values, 
+                                             int num_remote_values, 
+                                             void* context, 
+                                             int (*rhs_func)(void* context, real_t t, krylov_vector_t* x, krylov_vector_t* xdot),
+                                             int (*J_func)(void* context, real_t t, krylov_vector_t* x, krylov_vector_t* rhs, krylov_matrix_t* J),
+                                             void (*dtor)(void* context),
+                                             krylov_solver_t* solver,
+                                             krylov_pc_t* preconditioner,
+                                             krylov_matrix_t* matrix,
+                                             krylov_vector_t* vector)
+{
+  ASSERT(order >= 1);
+  ASSERT(order <= 5);
+  ASSERT(num_local_values > 0);
+  ASSERT(num_remote_values >= 0);
+  ASSERT(rhs_func != NULL);
+  ASSERT(J_func != NULL);
+  ASSERT(solver != NULL);
+  ASSERT(preconditioner != NULL);
+  ASSERT(matrix != NULL);
+  ASSERT(vector != NULL);
+
+  bdf_ode_t* integ = polymec_malloc(sizeof(bdf_ode_t));
+  integ->comm = comm;
+  integ->num_local_values = num_local_values;
+  integ->num_remote_values = num_remote_values;
+  integ->context = context;
+  integ->rhs = ink_rhs;
+  integ->k_rhs_func = rhs_func;
+  integ->k_J_func = J_func;
+  integ->k_solver = solver;
+  integ->k_precond = preconditioner;
+  integ->k_J = matrix;
+  integ->k_old_J = krylov_matrix_clone(matrix);
+  integ->k_x = vector;
+  integ->k_b = krylov_vector_clone(vector);
+  integ->dtor = dtor;
+  integ->status_message = NULL;
+  integ->t = 0.0;
+  integ->observers = ptr_array_new();
+  integ->error_weights = NULL;
+
+  integ->max_krylov_dim = 1;
+  integ->Jy = NULL;
+
+  // Set up KINSol and accessories.
+  integ->x = N_VNew(integ->comm, integ->num_local_values);
+  integ->x_with_ghosts = polymec_malloc(sizeof(real_t) * (integ->num_local_values + integ->num_remote_values));
+  integ->cvode = CVodeCreate(CV_BDF, CV_NEWTON);
+  CVodeSetMaxOrd(integ->cvode, order);
+  CVodeSetUserData(integ->cvode, integ);
+  CVodeInit(integ->cvode, bdf_evaluate_rhs, 0.0, integ->x);
+
+  // Set up the linear solvers.
+  // FIXME
+
+  ode_integrator_vtable vtable = {.step = bdf_step, .advance = bdf_advance, .reset = bdf_reset, .dtor = bdf_dtor};
+  char name[1024];
+  snprintf(name, 1024, "INK Backwards-Difference-Formulae (order %d)", order);
+  ode_integrator_t* I = ode_integrator_new(name, integ, vtable, order,
+                                           num_local_values + num_remote_values);
+
+  // Set default tolerances.
+  // relative error of 1e-4 means errors are controlled to 0.01%.
+  // absolute error is set to 1 because it's completely problem dependent.
+  bdf_ode_integrator_set_tolerances(I, 1e-4, 1.0);
+
+  return I;
 }
 

--- a/integrators/bdf_ode_integrator.h
+++ b/integrators/bdf_ode_integrator.h
@@ -8,6 +8,7 @@
 #ifndef POLYMEC_BDF_ODE_INTEGRATOR_H
 #define POLYMEC_BDF_ODE_INTEGRATOR_H
 
+#include "core/krylov_solver.h"
 #include "integrators/ode_integrator.h"
 #include "integrators/newton_pc.h"
 
@@ -52,6 +53,25 @@ ode_integrator_t* jfnk_bdf_ode_integrator_new(int order,
                                               newton_pc_t* precond,
                                               jfnk_bdf_krylov_t solver_type,
                                               int max_krylov_dim);
+
+// This function creates a BDF integrator that uses an inexact Newton-Krylov 
+// method to solve the underlying linearized equations. This method constructs 
+// a full Jacobian matrix and updates it only when needed, and requires a 
+// function to be specified for the update. The given Krylov solver, preconditioner,
+// matrix and vector objects will be used to conduct the solve and create 
+// workspaces, etc. The integrator assumes control of all these objects.
+ode_integrator_t* ink_bdf_ode_integrator_new(int order, 
+                                             MPI_Comm comm,
+                                             int num_local_values, 
+                                             int num_remote_values, 
+                                             void* context, 
+                                             int (*rhs_func)(void* context, real_t t, krylov_vector_t* x, krylov_vector_t* xdot),
+                                             int (*J_func)(void* context, real_t t, krylov_vector_t* x, krylov_vector_t* rhs, krylov_matrix_t* J),
+                                             void (*dtor)(void* context),
+                                             krylov_solver_t* solver,
+                                             krylov_pc_t* preconditioner,
+                                             krylov_matrix_t* matrix,
+                                             krylov_vector_t* vector);
 
 // This returns the context pointer passed to the bdf_ode_integrator 
 // constructor. In general, this will NOT return the same pointer as 

